File: connector.py
#!/usr/bin/env python3
import os, argparse, requests, sys

PROXY_ADDR = "http://localhost:9999"

parser = argparse.ArgumentParser(
    description="Adds a server to the proxy server list and runs the server"
)
parser.add_argument(
    "--host",
    help="host for server (ex. localhost:8888 or ex.com"
)
parser.add_argument(
    "--route",
    help="base route/slug (ex. .com/x/s route is /x/"
)
parser.add_argument(
    "--prog",
    help="program command to be run"
)
parser.add_argument(
    "-r",
    "--remove",
    help="remove the given host and route from the proxy",
    action="store_true"
)
args = parser.parse_args()

if not args.host:
    print("Must provide hostname")
    sys.exit(1)
if not args.route:
    print("Must provide route")
    sys.exit(1)

if args.host[-1] == "/": args.host = args.host[:-1]
if args.route[0] != "/": args.route = "/" + args.route
if args.route[-1] != "/": args.route += "/"

print(args.host, args.route, args.prog, args.remove)
sys.exit(0)

payload = {
    "host": args.host,
    "route": args.route,
    "connect": "1" if action.remove else "0"
}
try: req = requests.get(PROXY_ADDR, params=payload)
except:
    print(f"Error connecting to proxy (Proxy address: {PROXY_ADDR}")
    sys.exit(1)

if args.remove: sys.exit(0)
if args.prog:
    os.system(args.prog)
    payload["connect"] = "0"
    try: requests.get(PROXY_ADDR, param="payload")
    except: pass
else: print("Thirty seconds before server add to proxy")

File: connector.go
package main

import (
  "flag"
  "io/ioutil"
  "net/http"
  "net/url"
  // "os/exec"
  // "strings"
)

var (
  proxyURL string = "http://localhost:9999/"
)

func main() {
  // Get the flags from the command line
  // If adding a host, "route" must be combined with "host"
  // If removing a host, "remove" must be added as well as "route" or "host"
  route := flag.String("route", "", "The route for the proxy")
  host := flag.String("host", "", "The host/address of server")
  remove := flag.Bool("remove", false, "Remove the host or route given using the -host or -route flags")

  flag.Parse()
  var form url.Values
  if *remove {
    if *route != "" {
      form = url.Values{"remove": {"1"}, "route": {*route}}
    } else if *host != "" {
      form = url.Values{"remove": {"1"}, "host": {*host}}
    } else {
      println("Must provide host or route")
      return
    }
    resp, err := http.PostForm(proxyURL, form)
    if err != nil {
      println(err.Error())
      return
    }
    if body, err := ioutil.ReadAll(resp.Body); err != nil {
      println(err.Error())
    } else {
      println(string(body))
    }
    return
  } else if *route == "" && *host == "" {
    resp, err := http.Get(proxyURL)
    if err != nil {
      println(err.Error())
      return
    }
    if body, err := ioutil.ReadAll(resp.Body); err != nil {
      println(err.Error())
    } else {
      println(string(body))
    }
    return
  } else if *route == "" {
    println("Must provide route")
    return
  } else if *host == "" {
    println("Must provide host")
    return
  }
  // Clean the input
  if (*host)[len(*host)-1] == '/' {
  }
  if (*route)[0] != '/' {
    *route = "/" + *route
  }
  if (*route)[len(*route)-1] != '/' {
    *route += "/"
  }

  form := url.Values{"route": {*route}, "host": {*host}}
  resp, err := http.PostForm(proxyURL, form)
  if err != nil {
    println(err.Error())
    return
  }
  body, err := ioutil.ReadAll(resp.Body)
  if err != nil {
    println(err.Error())
  } else {
    println(string(body))
  }
}

func sendRequst(form url.Values) (string, error) {
  var err error
  var resp *http.Response
  if form == nil {
    resp, err = http.Get(proxyURL)
  } else {
    resp, err = http.PostForm(proxyURL, form)
  }
  if err != nil {
    return "", err
  }
  body, err := ioutil.ReadAll(resp.Body)
  return string(body), err
}

